Meta/Lagom/Tools/ELFBaker/main.cpp

```diff
@@ -0,0 +1,235 @@
+        if (first_program_header.p_filesz > (m_input_header.e_phoff + m_input_header.e_phentsize * m_input_header.e_phnum))
+            return Error::from_string_literal("First program header does not contain program header, use PHDRS flag in linker script");
+
+        return {};
+    }
+
+    size_t execute_program_headers(ByteBuffer& output)
+    {
+        T output_image(output);
+
+        for (auto& program_header : m_input_program_headers) {
+            if (program_header.p_type == PT_LOAD) {
+                auto src_span = m_input_data.span().slice(program_header.p_offset, program_header.p_filesz);
+                auto dst_span = output.span().slice(program_header.p_vaddr, program_header.p_filesz);
+                src_span.copy_to(dst_span);
+            }
+        }
+
+        for (auto& program_header : output_image.get_program_headers()) {
+            program_header.p_filesz = program_header.p_memsz;
+            program_header.p_offset = program_header.p_vaddr;
+        }
+
+        return m_program_loads_size;
+    }
+
+    size_t move_section_table(ByteBuffer& output, size_t offset)
+    {
+        T output_image(output);
+        auto& output_header = output_image.get_header();
+        size_t section_table_size = m_input_header.e_shnum * m_input_header.e_shentsize;
+
+        auto src_span = m_input_data.span().slice(m_input_header.e_shoff, section_table_size);
+        auto dst_span = output.span().slice(offset, section_table_size);
+        src_span.copy_to(dst_span);
+
+        output_header.e_shoff = offset;
+
+        return section_table_size;
+    }
+
+    size_t move_sections(ByteBuffer& output, size_t offset)
+    {
+        T output_image(output);
+
+        for (auto& section : output_image.get_sections()) {
+            if (section.sh_addr == 0 && section.sh_size > 0) {
+                auto src_span = m_input_data.span().slice(section.sh_offset, section.sh_size);
+                auto dst_span = output.span().slice(offset, section.sh_size);
+                src_span.copy_to(dst_span);
+
+                section.sh_offset = offset;
+
+                offset += section.sh_size;
+            } else {
+                section.sh_offset = section.sh_addr;
+            }
+        }
+
+        return offset;
+    }
+
+    size_t compute_program_load_sizes() const
+    {
+        size_t total = 0;
+
+        for (auto& program_header : m_input_program_headers) {
+            size_t program_end = program_header.p_vaddr + program_header.p_memsz;
+            total = max(total, program_end);
+        }
+
+        return total;
+    }
+
+    size_t compute_relocated_sections_size() const
+    {
+        size_t total = 0;
+
+        for (auto& section : m_input_sections) {
+            if (section.sh_addr == 0 && section.sh_size > 0) {
+                total += section.sh_size;
+            }
+        }
+
+        return total;
+    }
+
+    ByteBuffer& m_input_data;
```

Comment:
Does the input really want to be mutable? Inputs are normally immutable and outputs mutable(?)